### PR TITLE
 rgw: local runner for rgw tasks

### PR DESCRIPTION
**Pull request by oritwas**
run rgw tasks against a local cluster (run by vstart)

### CODE REVIEW

**Comment by cbodley** [comment]
very cool! would it be possible to include a script+conifg in qa/workunits/rgw/ that can run the local radosgw_admin or s3test task? a working example would make it a lot easier to get it running for other stuff

File: qa/tasks/local_rgw.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    client = clients_from_config[0]
+    log.debug('client is: %r', client)
+
+    #role_endpoints = assign_ports(ctx, config)
+    #ctx.rgw.role_endpoints = role_endpoints
+    ctx.rgw.use_fastcgi = True
+    ctx.rgw.frontend = config.pop('frontend', 'civetweb')
+
+    ctx.rgw.role_endpoints = {}
+    ctx.rgw.role_endpoints[client] = ('localhost', 8000)
```
Comment (cbodley) [inline qa/tasks/local_rgw.py]: consider getting the port number from config - that would make it easy to configure multiple gateways too, ie:
```yaml
local_rgw:
  client.rgw.0: 8000
  client.rgw.1: 8001
```

File: qa/tasks/local_rgw.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import argparse
+import contextlib
+import logging
+
+from teuthology import misc as teuthology
+
+log = logging.getLogger(__name__)
+
+def assign_ports(ctx, config):
```
Comment (cbodley) [inline qa/tasks/local_rgw.py]: unused here

**Comment by mattbenjamin** [comment]
@oritwas yay!

**Comment by alimaredia** [comment]
This should be good to merge once https://github.com/ceph/ceph/pull/22538 is merged. Do you want me to write up any documentation?

**Comment by alimaredia** [comment]
Here are my initial notes on how to run s3 tests with this branch:

https://pad.ceph.com/p/vstart_teuthology_runner_notes

Let me know if there are any issues you run into.

**Comment by gregsfortytwo** [comment]
@alimaredia are you just waiting for somebody to confirm those instructions work?

**Comment by alimaredia** [comment]
@gregsfortytwo There are two more things I need to do:
- Add docs to this (the instructions work)
- Look into why the output from the s3-tests task is going to std_out.

Hope to wrap both of those up by the next testing weekly meeting.

**Comment by gregsfortytwo** [comment]
Sweet, thanks!

**Comment by tianshan** [comment]
@alimaredia  there is a little issue to run task radosgw-admin,  the ceph-coverage and radosgw-admin cmd wrapper rgwadmin() in  qa/tasks/util/rgw.py need {tdir} perfix, and the ceph-coverage need execute permission

**Comment by tianshan** [comment]
@alimaredia sorry it's may fault that I have another version radosgw-admin installed.  after put the cluster_bin_dir in front of os path, it works!

File: qa/tasks/local_cluster.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        rgw_enable_usage_log = true
+      - everything is a very, very early PoC.
+    """
+    # we need to update the PATH variable to let shell find binaries
+    # of a cluster deployed with vstart.sh
+    try:
+      cluster_bin_dir = config['cluster_path'] + os.sep + 'bin'
+    except TypeError:
+        raise ValueError("the mandatory cluster_path is undefined")
+    os.environ['PATH'] += os.pathsep + cluster_bin_dir
```
Comment (tianshan) [inline qa/tasks/local_cluster.py]: put cluster_bin_dir in front of os.environ['PATH'] is better

**Comment by mattbenjamin** [comment]
unstale me, my man

**Comment by alimaredia** [comment]
unstale this

**Comment by mattbenjamin** [comment]
Reanimate

**Comment by yuvalif** [comment]
@alimaredia, this is great! going to try it out.

**Pull request closed without merging**